numpy/lib/histograms.py

```diff
@@ -97,6 +97,39 @@ def _hist_bin_scott(x):
     return (24.0 * np.pi**0.5 / x.size)**(1.0 / 3.0) * np.std(x)
 
 
+def _hist_bin_ise(x):
+    """
+    Histogram bin estimator based on minimizing the estimated integrated squared error (ISE).
+
+    The number of bins is chosen by minimizing the estimated ISE against the unknown true distribution.
+    The ISE is estimated using cross-validation and can be regarded as a generalization of Scott's rule.
+    https://en.wikipedia.org/wiki/Histogram#Scott.27s_normal_reference_rule
+
+    Parameters
+    ----------
+    x : array_like
+        Input data that is to be histogrammed, trimmed to range. May not
+        be empty.
+
+    Returns
+    -------
+    h : An estimate of the optimal bin width for the given data.
+    """
+
+    n = x.size
+    ptp_x = np.ptp(x)
+    if n <= 1 or ptp_x == 0:
+        return 0
+
+    def jhat(nbins):
+        hh = ptp_x / nbins
+        p_k = np.histogram(x, bins=nbins)[0] / n
```

Comment:
I think this maybe needs to inherit the `range` argument from the level above? Otherwise it's not actually using the optimal bins.

Reply:
Do you mean
`p_k = np.histogram(x, bins=nbins, range=range)[0] / n`
for that line?
But the number of bins is fixed at `nbins`, so why do we need `range`?

Reply:
`range` changes where the bins end up. If the top level call to `histogram` uses `range`, but this call doesn't, then you won't actually be producing the optimal result.

Reply:
`_hist_bin_*` functions returns binswidth, not bin number and from my tests below, if `range` covers all the data, the binswidth is not affected.
```
import numpy as np
import numpy.random
x = np.random.normal(loc=0, scale=1, size=1000)
np.unique(np.diff(np.histogram(x, bins='ise')[1]))
Out[5]: array([0.34771668, 0.34771668, 0.34771668, 0.34771668, 0.34771668])
np.unique(np.diff(np.histogram(x, bins='ise', range=(-1e6, 1e6))[1]))
Out[6]: array([0.34771663, 0.34771663, 0.34771663, 0.34771663])
```

Reply:
The issue is when the  range does not cover all the data. The bins are free to shift side-to-side, and as a result the integrated squared error is not what you expected it to be.

Reply:
done